File: rle.go
// rle.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

// ANSI colors
const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
)

func colorize(text, color string) string {
	return color + text + reset
}

const ESCAPE = '\\'

func compress(text string) string {
	if text == "" {
		return ""
	}
	var result strings.Builder
	n := len(text)
	i := 0
	for i < n {
		ch := text[i]
		j := i + 1
		for j < n && text[j] == ch {
			j++
		}
		count := j - i
		if count >= 3 {
			result.WriteByte(ch)
			result.WriteByte(ESCAPE)
			result.WriteString(strconv.Itoa(count))
		} else {
			for k := 0; k < count; k++ {
				if ch == ESCAPE {
					result.WriteByte(ESCAPE)
				}
				result.WriteByte(ch)
			}
		}
		i = j
	}
	return result.String()
}

func decompress(text string) (string, error) {
	if text == "" {
		return "", nil
	}
	var result strings.Builder
	n := len(text)
	i := 0
	for i < n {
		ch := text[i]
		if ch == ESCAPE {
			// Проверяем, экранированный ли это escape
			if i+1 < n && text[i+1] == ESCAPE {
				result.WriteByte(ESCAPE)
				i += 2
				continue
			}
			// Управляющая последовательность
			if i+1 >= n {
				return "", errors.New("неожиданный конец строки после escape")
			}
			char := text[i+1]
			i += 2
			// Читаем число
			numStr := ""
			for i < n && text[i] >= '0' && text[i] <= '9' {
				numStr += string(text[i])
				i++
			}
			if numStr == "" {
				return "", errors.New("отсутствует число после escape")
			}
			count, err := strconv.Atoi(numStr)
			if err != nil {
				return "", err
			}
			for k := 0; k < count; k++ {
				result.WriteByte(char)
			}
		} else {
			result.WriteByte(ch)
			i++
		}
	}
	return result.String(), nil
}

func readInput(filename string) (string, error) {
	if filename == "-" || filename == "" {
		// читаем stdin
		reader := bufio.NewReader(os.Stdin)
		var builder strings.Builder
		for {
			line, err := reader.ReadString('\n')
			if err != nil && err != io.EOF {
				return "", err
			}
			builder.WriteString(line)
			if err == io.EOF {
				break
			}
		}
		return builder.String(), nil
	}
	data, err := os.ReadFile(filename)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func writeOutput(filename string, content string) error {
	if filename == "-" || filename == "" {
		fmt.Print(content)
		return nil
	}
	return os.WriteFile(filename, []byte(content), 0644)
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println(colorize("Usage: rle compress|decompress [input] [output]", yellow))
		os.Exit(1)
	}
	mode := os.Args[1]
	inputFile := ""
	outputFile := ""
	if len(os.Args) >= 3 {
		inputFile = os.Args[2]
	}
	if len(os.Args) >= 4 {
		outputFile = os.Args[3]
	}
	if mode != "compress" && mode != "decompress" {
		fmt.Println(colorize("Invalid mode. Use compress or decompress.", red))
		os.Exit(1)
	}

	// Чтение
	data, err := readInput(inputFile)
	if err != nil {
		fmt.Println(colorize("Error reading input: "+err.Error(), red))
		os.Exit(1)
	}

	var result string
	var err2 error
	if mode == "compress" {
		result = compress(data)
	} else {
		result, err2 = decompress(data)
		if err2 != nil {
			fmt.Println(colorize("Decompression error: "+err2.Error(), red))
			os.Exit(1)
		}
	}

	if err := writeOutput(outputFile, result); err != nil {
		fmt.Println(colorize("Error writing output: "+err.Error(), red))
		os.Exit(1)
	}
	if outputFile != "" && outputFile != "-" {
		fmt.Println(colorize("Result written to "+outputFile, green))
	}
}

File: rle.cpp
// rle.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <cctype>
#include <stdexcept>

using namespace std;

const string RESET = "\033[0m";
const string GREEN = "\033[92m";
const string RED = "\033[91m";
const string YELLOW = "\033[93m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

const char ESCAPE = '\\';

string compress(const string& text) {
    if (text.empty()) return "";
    ostringstream result;
    size_t i = 0, n = text.size();
    while (i < n) {
        char ch = text[i];
        size_t j = i + 1;
        while (j < n && text[j] == ch) ++j;
        size_t count = j - i;
        if (count >= 3) {
            result << ch << ESCAPE << count;
        } else {
            for (size_t k = 0; k < count; ++k) {
                if (ch == ESCAPE) result << ESCAPE;
                result << ch;
            }
        }
        i = j;
    }
    return result.str();
}

string decompress(const string& text) {
    if (text.empty()) return "";
    ostringstream result;
    size_t i = 0, n = text.size();
    while (i < n) {
        char ch = text[i];
        if (ch == ESCAPE) {
            if (i + 1 < n && text[i + 1] == ESCAPE) {
                result << ESCAPE;
                i += 2;
                continue;
            }
            if (i + 1 >= n) throw runtime_error("Unexpected end after escape");
            char repeat_char = text[i + 1];
            i += 2;
            string num_str;
            while (i < n && isdigit(text[i])) {
                num_str += text[i];
                ++i;
            }
            if (num_str.empty()) throw runtime_error("Missing number after escape");
            int count = stoi(num_str);
            for (int k = 0; k < count; ++k) result << repeat_char;
        } else {
            result << ch;
            ++i;
        }
    }
    return result.str();
}

string readFile(const string& filename) {
    if (filename == "-" || filename.empty()) {
        // read stdin
        stringstream buffer;
        buffer << cin.rdbuf();
        return buffer.str();
    }
    ifstream f(filename);
    if (!f) throw runtime_error("Cannot open file: " + filename);
    stringstream buffer;
    buffer << f.rdbuf();
    return buffer.str();
}

void writeFile(const string& filename, const string& content) {
    if (filename == "-" || filename.empty()) {
        cout << content;
        return;
    }
    ofstream f(filename);
    if (!f) throw runtime_error("Cannot write file: " + filename);
    f << content;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cout << colorize("Usage: rle compress|decompress [input] [output]", YELLOW) << endl;
        return 1;
    }
    string mode = argv[1];
    if (mode != "compress" && mode != "decompress") {
        cout << colorize("Invalid mode. Use compress or decompress.", RED) << endl;
        return 1;
    }
    string inputFile = (argc >= 3) ? argv[2] : "";
    string outputFile = (argc >= 4) ? argv[3] : "";

    string data;
    try {
        data = readFile(inputFile);
    } catch (const exception& e) {
        cout << colorize("Error reading input: " + string(e.what()), RED) << endl;
        return 1;
    }

    string result;
    try {
        if (mode == "compress") result = compress(data);
        else result = decompress(data);
    } catch (const exception& e) {
        cout << colorize("Error: " + string(e.what()), RED) << endl;
        return 1;
    }

    try {
        writeFile(outputFile, result);
        if (!outputFile.empty() && outputFile != "-") {
            cout << colorize("Result written to " + outputFile, GREEN) << endl;
        }
    } catch (const exception& e) {
        cout << colorize("Error writing output: " + string(e.what()), RED) << endl;
        return 1;
    }
    return 0;
}
